src/qililab/qprogram/qblox_compiler.py

```diff
@@ -328,118 +333,137 @@ def _handle_set_offset(self, element: SetOffset):
-            else convert(element.time)
+        duration: QPyProgram.Register | int
+        if isinstance(element.duration, Variable):
+            duration = self._buses[element.bus].variable_to_register[element.duration]
+            self._buses[element.bus].dynamic_durations.append(element.duration)
+        else:
+            convert = QbloxCompiler._convert_value(element)
+            duration = convert(element.duration)
+            self._buses[element.bus].static_duration += duration
+        self._buses[element.bus].qpy_block_stack[-1].append_component(
+            component=QPyInstructions.Wait(wait_time=duration)
         )
-        self._buses[element.bus].qpy_block_stack[-1].append_component(component=QPyInstructions.Wait(wait_time=time))
+        self._buses[element.bus].marked_for_sync = True
 
-    def _handle_sync(self, _: Sync):
-        # QBlox does not currently support syncing on selective sequences. If it did we would do somethin like:
-        # buses = element.buses if element.buses is not None else self._buses.keys()
-        for bus in self._buses:
-            self._buses[bus].qpy_block_stack[-1].append_component(
-                component=QPyInstructions.WaitSync(wait_time=QbloxCompiler.minimum_wait_duration)
-            )
+    def _handle_sync(self, element: Sync):
+        # Get the buses involved in the sync operation.
+        buses = set(element.buses or self._buses)
+
+        # If they are zero or one, return.
+        if len(buses) <= 1:
+            return
+
+        # If there is no bus marked for sync, return.
+        if all(not self._buses[bus].marked_for_sync for bus in buses):
+            return
+
+        # Is there any bus that has dynamic durations?
+        if any(bus for bus in buses if self._buses[bus].dynamic_durations or self._buses[bus].sync_durations):
+            # If yes, we must add a sync block that calculates the difference between buses dynamically.
+            # But the following doesn't work unfortunetely, so raise an error for now.
+            self.__handle_dynamic_sync(buses=buses)
+        else:
+            # If no, calculating the difference is trivial.
+            self.__handle_static_sync(buses=buses)
+
+        # In any case, mark al buses as synced.
+        for bus in buses:
+            self._buses[bus].marked_for_sync = False
+
+    def __handle_static_sync(self, buses: set[str]):
+        max_duration = max(self._buses[bus].static_duration for bus in buses)
+        for bus in buses:
+            duration_diff = max_duration - self._buses[bus].static_duration
+            if duration_diff > 0:
+                self._buses[bus].qpy_block_stack[-1].append_component(
+                    component=QPyInstructions.Wait(wait_time=duration_diff)
+                )
+                self._buses[bus].static_duration += duration_diff
+
+    def __handle_dynamic_sync(self, buses: set[str]):
+        raise NotImplementedError("Dynamic syncing is not implemented yet.")
 
```

Comment:
the recommendation from the QBLOX guys to handle times / sync between sequencers without running `wait_sync` (in case this helps) was to just copy the Q1ASM string from sequencers being used and substitute the `play` instructions by `wait`.
Maybe something similar could be done here to avoid keeping track of times all the time. For example setting a play instruction with a specific index that we know it's a dummy play instruction and parse that to a wait before uploading (though that feels very hacky, but maybe something along those lines)

Reply:
I don't understand what the recommendation is.

This method had originally around ~120 lines of code to calculate the timings, but unfortunetely the hardware does not do some of the calculations correctly, so the code was replaced with an error for now.

The only other way to implement this is to "unwrap" the loops, as in c/p all operations within, with pre-calculated values.